### ISSUE TITLE
Support for vector tiles hosted on ArcGIS Enterprise

### ISSUE THREAD

**Issue by roemhildtg** [Esri/esri-leaflet-vector#28]
I couldn't find any way to add a url to a custom basemap resource. Is this possible, or are only the default esri basemaps supported?

**Comment by jgravois** [comment]
because multiple styles can be defined and associated with a single cached tile store, this plugin expects custom basemaps to be referenced via their item id.
```js
// http://www.arcgis.com/home/item.html?id=0bac0ffdc8634d9a9bc662bb8fa7547d
L.esri.Vector.layer('0bac0ffdc8634d9a9bc662bb8fa7547d').addTo(map);
```
[(ref)](https://github.com/Esri/esri-leaflet-vector/blob/35a720755987d0ee3e8801a93caa1edb1aaaf4f8/debug/sample.html#L39-L41)

**Comment by roemhildtg** [comment]
@jgravois thanks for the response. So only vector tile layers hosted on AGO are supported? 
Are there plans to support resources on local ArcGIS Servers/Portals?

**Comment by jgravois** [comment]
it wouldn't be very difficult to add some sort of configuration option to allow folks to specify an alternate portal url.

to be honest, i'm juggling lots of other projects so if you (or anyone else that's interested) could provide urls for public test services and some more detailed information about what exactly you'd like to see, progress would be a lot more likely.

**Comment by roemhildtg** [comment]
Totally understandable. I think this would be the approach I would be using:

```javascript
L.esri.Vector.layer('https://www.arcgis.com/sharing/rest/content/items/bdf1eec3fa79456c8c7c2bb62f86dade/resources/styles/root.json').addTo(map)
```

The js api supports these types of urls:
```javascript
// Typical usage
var vtlLayer = new VectorTileLayer({
  // URL to the style of vector tiles
  url: "https://www.arcgis.com/sharing/rest/content/items/bf79e422e9454565ae0cbe9553cf6471/resources/styles/root.json"
});

var vtlLayer = new VectorTileLayer({
  // URL to the vector tile service
  url: "https://basemaps.arcgis.com/b2/arcgis/rest/services/World_Basemap/VectorTileServer"
});
```

Thoughts on that? I could look into making a pull request with the functionality.

**Comment by jgravois** [comment]
to be honest, i don't really *love* the way the JSAPI expects developers to supply input parameters for `VectorTileLayer`.

it seems an unnecessary burden for the majority of developers that use ArcGIS Online to have to construct an obscure url for the sake of the minority that use a custom portal.

i haven't given it much thought, but my first instinct is that we do something like this:

```js
// optional
L.esri.Vector.options.portal = 'https://www.arcgis.com/sharing/';

L.esri.Vector.layer('bf79e422e9454565ae0cbe9553cf6471').addTo(map);
```

**Comment by roemhildtg** [comment]
Not trying to be difficult, but from my perspective I might like to use _both_ arcgisonline basemaps _and_ a portal basemap. So wouldn't configuring the portal url globally would prevent that?

**Comment by jgravois** [comment]
> Not trying to be difficult

no need to apologize.  i hadn't considered that at all.

we could just as easily pass the information along in an *optional* constructor option.

```js
L.esri.Vector.layer('bf79e422e9454565ae0cbe9553cf6471', {
  portal: 'https://custom.com/arcgis/sharing/'
}).addTo(map);
```
the plugin is experimental at this point, so i'm not against making a breaking change.  i just want to make sure that whatever we decide on going forward prioritizes simplicity for the most common use cases.

**Comment by roemhildtg** [comment]
IMO, it would also be nice to have the flexibility of using the actual server endpoint rather than requiring the portal url/item-id. 

`https://basemaps.arcgis.com/b2/arcgis/rest/services/World_Basemap/VectorTileServer`

I'm not sure, but is there a major benefit to using the item id/portal url? Does the portal json give esri-leaflet more information?

**Comment by jgravois** [comment]
my understanding is that the tile server itself will have an associated default style so there'd certainly be no harm in accepting those urls.

my initial approach in this plugin required focusing on style `ids` because all the Esri vector basemaps point at the exact same base data.  in practice, i don't know how come it is/will be for customers to create more than one style for their own basemaps.

**Comment by roemhildtg** [comment]
From my perspective I doubt I'd ever have more than one style per basemap. This is because it seems much easier to publish a basemap from ArcGIS Pro than to go through and modify all the styles by hand. I've already published a couple of basemaps using nearly the same data and the size is incredibly small compared to regular tile layers so having more than one copy of the data is not a big deal.

**Comment by gavinr** [comment]
So just to get this discussion rolling again on this ... Would these four test cases be what we want to support going forward?

```
// 1. named basemap (existing)
L.esri.Vector.basemap('Nova').addTo(map);

// 2. custom hosted basemap (existing)
L.esri.Vector.layer('bd505ce3efff479bb4e87b182f180159').addTo(map);

// 3. Enterprise - by id (new)
L.esri.Vector.layer({
  id: 'bf79e422e9454565ae0cbe9553cf6471',
  portal: 'https://custom.com/arcgis/'
}).addTo(map);

// 4. by URL (new)
L.esri.Vector.layer({
  url: 'https://basemaps.arcgis.com/b2/arcgis/rest/services/World_Basemap/VectorTileServer'
}).addTo(map);

**Comment by COV-GIS** [comment]
Number 5 use case would be to consume an unhosted custom `.json` file directly.

```js
L.esri.Vector.layer({
  url: 'https://www.example.com/hybrid-reference.json'
}).addTo(map);
```

@gavinr I see you're assigned. Any plans to work on this? Or in the works?

**Comment by roemhildtg** [comment]
@COV-GIS I do believe that 4 would handle that also, so long as you've set up iis to host the json file as a default document.

**Comment by COV-GIS** [comment]
I guess when I see the url in no. 4 I'm thinking vtl source and not style like `root.json` or what have you. Methinks it should be a style json endpoint or file as the source(s) would be included therein.

**Comment by roemhildtg** [comment]
Ah I see. Yeah that would be yet a different case.

**Comment by gavinr** [comment]
@COV-GIS thanks for the additional discussions. I did take a look recently but it turned out to be more refactoring than i had time for at the time, so I had to shelve it. It's still on the back of my mind, but if you'd like to take a whack at it, you're definitely welcome.

**Comment by COV-GIS** [comment]
Thanks @gavinr. I might take a look. `esri` + `leaflet` looks promising for simple maps. But I want my custom VTLs.

**Comment by irzu** [comment]
Could you possibly advise whether a layer requiring a token can be used?
tried `L.esri.Vector.layer('12345?token=...').addTo(map);` - didn't work

**Comment by jwasilgeo** [comment]
We hope to add support for this when we release a new version of this library, [following the current v3.0.1](https://github.com/esri/esri-leaflet-vector/compare/v3.0.1...HEAD).

**Comment by jwasilgeo** [comment]
This is available now in the recent [v3.1.0 release](https://github.com/Esri/esri-leaflet-vector/releases/tag/v3.1.0). 🎉
